pkg/ingester/ingester_v2.go

```diff
@@ -190,18 +204,45 @@ func (i *Ingester) v2Push(ctx old_ctx.Context, req *client.WriteRequest) (*clien
 	// successfully committed
 	succeededSamplesCount := 0
 	failedSamplesCount := 0
+	now := model.Now().UnixNano()
 
 	// Walk the samples, appending them to the users database
 	app := db.Appender()
 	for _, ts := range req.Timeseries {
-		// Convert labels to the type expected by TSDB
-		lset := client.FromLabelAdaptersToLabelsWithCopy(ts.Labels)
+		// Check if we already have a cached reference for this series. Be aware
+		// that even if we have a reference it's not guaranteed to be still valid.
+		cachedRef, cachedRefExists := db.refCache.GetRef(now, client.FromLabelAdaptersToLabels(ts.Labels))
 
 		for _, s := range ts.Samples {
-			_, err := app.Add(lset, s.TimestampMs, s.Value)
-			if err == nil {
-				succeededSamplesCount++
-				continue
+			var err error
+
+			// If the cached reference exists, we try to use it.
+			if cachedRefExists {
+				if err = app.AddFast(cachedRef, s.TimestampMs, s.Value); err == nil {
+					succeededSamplesCount++
+					continue
+				}
+
+				if err == tsdb.ErrNotFound {
+					cachedRefExists = false
+				}
```

Comment:
Would be safe to set `err` to `nil` inside here to avoid any subtle bugs in the future.